File: Tecnicas_de_prog_2/epl-scala/src/main/scala/br/unb/cic/epl/Mult.scala
package br.unb.cic.epl

package object Mult {
  trait GMult extends Core.Expression {
    type T <: Core.Expression           //abstract type in Scala
    var lhs: T = _ 
    var rhs: T = _ 
    override def print(): String = "(" + lhs.print() + " * " + rhs.print() + ")" 
  }

  class Mult(l: Core.Expression, r: Core.Expression) extends GMult {
    type T = Core.Expression

    lhs = l
    rhs = r
  }
}

File: Tecnicas_de_prog_2/epl/src/br/unb/cic/tp2/epl/Mult.java
package br.unb.cic.tp2.epl;

/**
 * An expression for representing Multiplications
 * @author potPotato
 */
public class Mult implements Exp {

	private Exp lhs; 
	private Exp rhs; 
	
	public Mult(Exp lhs, Exp rhs) {
		this.lhs = lhs;
		this.rhs = rhs;
	}
	
	public int eval() {
		return lhs.eval() * rhs.eval();
	}
	public String print() {
		StringBuilder sb = new StringBuilder();
		sb.append("(");
		sb.append(lhs.print());
		sb.append(" * ");
		sb.append(rhs.print());
		sb.append(")");
		return sb.toString();
	}
}